src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -487,349 +495,297 @@ public Task<Compilation> GetCompilationAsync(SolutionCompilationState compilatio
 
                     var state = ReadState();
 
-                    return state switch
-                    {
-                        // if we already have a compilation, we must be already done!  This can happen if two
-                        // threads were waiting to build, and we came in after the other succeeded.
-                        FinalState finalState
-                            => finalState,
+                    if (state is FinalState finalState)
+                        return finalState;
```

Comment:
this is the main logic to go from whatever state we have to the final-compiled-state.   i really like how this has turned out.

Reply:
first, if we have the final state, we're done.